File: lib/src/main/java/by/mrrockka/domain/payout/Payout.java
package by.mrrockka.domain.payout;

import by.mrrockka.domain.Person;
import by.mrrockka.domain.collection.PersonBounties;
import by.mrrockka.domain.collection.PersonEntries;
import by.mrrockka.domain.collection.PersonWithdrawals;
import lombok.Builder;
import lombok.NonNull;

import java.math.BigDecimal;
import java.util.List;

@Builder(toBuilder = true)
public record Payout(@NonNull PersonEntries personEntries, PersonWithdrawals personWithdrawals,
                     PersonBounties personBounties, List<Payer> payers) {
  public BigDecimal total() {
    return payers().stream().map(Payer::amount).reduce(BigDecimal::add).orElse(BigDecimal.ZERO);
  }

  public Person person() {
    return personEntries().person();
  }
}

File: core/src/main/java/by/mrrockka/domain/Payout.kt
package by.mrrockka.domain

import java.math.BigDecimal

data class Payout<PLAYER : Player>(
        val player: PLAYER,
        val payers: List<Payer<PLAYER>>,
)

typealias TournamentPayout = Payout<TournamentPlayer>
typealias CashPayout = Payout<CashPlayer>
typealias BountyPayout = Payout<BountyPlayer>

fun Payout<*>.total(): BigDecimal {
    return this.payers.map { it.amount }.fold(BigDecimal.ZERO, BigDecimal::add)
}
